crates/edr_solidity_tests/src/runner.rs

```diff
@@ -95,42 +106,211 @@ impl<'a> ContractRunner<'a> {
         Self {
             name,
             contract,
-            executor,
             revert_decoder,
+            known_contracts,
+            contract_decoder,
             initial_balance,
             sender,
             test_fail,
             solidity_fuzz_fixtures,
+            executor_builder,
         }
     }
 }
 
-impl<'a> ContractRunner<'a> {
+impl<'a, NestedTraceDecoderT: SyncNestedTraceDecoder> ContractRunner<'a, NestedTraceDecoderT> {
+    /// Runs all tests for a contract whose names match the provided regular
+    /// expression
+    pub fn run_tests(
+        self,
+        filter: &dyn TestFilter,
+        test_options: &TestOptions,
+        handle: &tokio::runtime::Handle,
+    ) -> SuiteResult {
+        info!("starting tests");
+        let start = Instant::now();
+        let mut warnings = Vec::new();
+
+        let mut executor = self.executor_builder.clone().build();
+
+        let setup_fns: Vec<_> = self
+            .contract
+            .abi
+            .functions()
+            .filter(|func| func.name.is_setup())
+            .collect();
+
+        let needs_setup = setup_fns.len() == 1 && setup_fns[0].name == "setUp";
+
+        // There is a single miss-cased `setUp` function, so we add a warning
+        for &setup_fn in setup_fns.iter() {
+            if setup_fn.name != "setUp" {
+                warnings.push(format!(
+                    "Found invalid setup function \"{}\" did you mean \"setUp()\"?",
+                    setup_fn.signature()
+                ));
+            }
+        }
+
+        // There are multiple setUp function, so we return a single test result for
+        // `setUp`
+        if setup_fns.len() > 1 {
+            return SuiteResult::new(
+                start.elapsed(),
+                [(
+                    "setUp()".to_string(),
+                    TestResult::fail("multiple setUp functions".to_string()),
+                )]
+                .into(),
+                warnings,
+                self.contract.libraries.clone(),
+            );
+        }
+
+        let has_invariants = self
+            .contract
+            .abi
+            .functions()
+            .any(foundry_evm::abi::TestFunctionExt::is_invariant_test);
+
+        // Invariant testing requires tracing to figure out what contracts were created.
+        let tmp_tracing = executor.inspector.tracer.is_none() && has_invariants && needs_setup;
+        // let tmp_tracing = true;
```

Comment:
Leftover?

Reply:
Thanks, fixed in https://github.com/NomicFoundation/edr/pull/790/commits/6859268e37f49f97d142c846bfb6b94d4ef6c03f